upgrade/upgrade.go

```diff
@@ -148,61 +148,53 @@ func doAPIVersionUpgrade(namespace string, config *restclient.Config, lhClient *
 			return err
 		}
-		// v1alpha1 doesn't have a setting entry for crd-api-version, need to create
-		crdAPIVersionSetting = &longhorn.Setting{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: string(types.SettingNameCRDAPIVersion),
-			},
-			Value: "",
-		}
-		crdAPIVersionSetting, err = lhClient.LonghornV1beta1().Settings(namespace).Create(context.TODO(), crdAPIVersionSetting, metav1.CreateOptions{})
-		if err != nil {
-			return errors.Wrap(err, "cannot create CRDAPIVersionSetting")
-		}
 	} else {
 		crdAPIVersion = crdAPIVersionSetting.Value
 	}
 
 	if crdAPIVersion != "" &&
-		crdAPIVersion != types.CRDAPIVersionV1alpha1 &&
-		crdAPIVersion != types.CRDAPIVersionV1beta1 {
+		crdAPIVersion != types.CRDAPIVersionV1beta1 &&
+		crdAPIVersion != types.CRDAPIVersionV1beta2 {
 		return fmt.Errorf("unrecognized CRD API version %v", crdAPIVersion)
 	}
+
 	if crdAPIVersion == types.CurrentCRDAPIVersion {
 		logrus.Info("No API version upgrade is needed")
 		return nil
 	}
 
 	switch crdAPIVersion {
 	case "":
-		// it can be a new installation or v1alpha1
-		isV1alpha1, err := v1alpha1.IsCRDVersionMatch(config, namespace)
-		if err != nil {
-			logrus.Warnf("Cannot verify current CRD version, assume it's not v1alpha1: %v", err)
+		// upgradable: supported version or new installation ((v1beta1, v1beta2)
+		// non-upgradable: non-supported version (v1alpha1 which cannot upgrade directly)
+		upgradable, err := v1beta1.CanUpgrade(config, namespace)
+		if !upgradable {
+			return err
 		}
-		if !isV1alpha1 {
-			crdAPIVersionSetting.Value = types.CurrentCRDAPIVersion
-			if _, err := lhClient.LonghornV1beta1().Settings(namespace).Update(context.TODO(), crdAPIVersionSetting, metav1.UpdateOptions{}); err != nil {
-				return errors.Wrapf(err, "cannot finish CRD API upgrade by setting the CRDAPIVersionSetting to %v", types.CurrentCRDAPIVersion)
-			}
-			logrus.Infof("Initialized CRD API Version to %v", types.CurrentCRDAPIVersion)
-			break
+
+		crdAPIVersionSetting = &longhorn.Setting{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: string(types.SettingNameCRDAPIVersion),
+			},
+			Value: types.CurrentCRDAPIVersion,
 		}
-		fallthrough
-	case types.CRDAPIVersionV1alpha1:
-		if types.CurrentCRDAPIVersion != types.CRDAPIVersionV1beta1 {
-			return fmt.Errorf("cannot upgrade from %v to %v directly", types.CRDAPIVersionV1alpha1, types.CurrentCRDAPIVersion)
+		_, err = lhClient.LonghornV1beta2().Settings(namespace).Create(context.TODO(), crdAPIVersionSetting, metav1.CreateOptions{})
+		if err != nil {
+			return errors.Wrap(err, "cannot create CRDAPIVersionSetting")
 		}
-		if err := v1alpha1.UpgradeFromV1alpha1ToV1beta1(config, namespace, lhClient); err != nil {
+		logrus.Infof("New %v installation", types.CurrentCRDAPIVersion)
+	case types.CRDAPIVersionV1beta1:
+		logrus.Infof("Upgrading from %v to %v", types.CRDAPIVersionV1beta1, types.CurrentCRDAPIVersion)
+		if err := v1beta1.UpgradeFromV1beta1ToV1beta2(config, namespace, lhClient); err != nil {
```

Comment:
I'm thinking that this step should be put in like v123to130 package.

So, `doAPIVersionUpgrade` only upgrades the Setting `crd-api-version`.
And the corresponding conditions conversion from the map to slice should be into v123to130.
Also, we should copy existing v1beta1 CR into v1beta2 CR in v123to130 as well.

Therefore, we don't have to change the upgrade packages v070to080, v100to101, v102to110, v110to111, v110to120, v111to120, v120to121, v122to123.

In general, the package v123to130 should involve:
- Copy the existing v1beta1 CR to v1beta2 CR
- Convert the map to slice of the `status.conditions`

What do you and others think? @derekbit @innobead @shuo-wu @joshimoo

Reply:
I agree with it. It will be unified.

Reply:
But in upgrade functions, they uses the datastore and engine, which accepts v1beta2 objects now, but some objects are got by v1beta1 APIs. They are incompatible. Any idea to resolve?

Reply:
> But in upgrade functions, they uses the datastore and engine, which accepts v1beta2 objects now, but some objects are got by v1beta1 APIs. They are incompatible. Any idea to resolve?

No from my side. Let's keep your current way that directly upgrades to v1beta2. Otherwise, have an interface to converts the datastore <-> v1beta1/v1beta2 and engine <-> v1beta1/v1beta2 is too complex.

Reply:
> So, doAPIVersionUpgrade only upgrades the Setting crd-api-version.
And the corresponding conditions conversion from the map to slice should be in v123to130.
Also, we should copy existing v1beta1 CR into v1beta2 CR in v123to130 as well.

We need to upgrade v1beta1 CR to v1beta2  first because status.Conditions is changed, or we will encounter
```
time="2021-11-24T07:23:22Z" level=fatal msg="Error starting manager: upgrade CRD failed: upgrade from v0.7.0 to v0.8.0: upgrade instance manager failed: upgrade from v0.7.0 to v0.8.0: failed to list all nodes during the instance managers upgrade: v1beta2.NodeList.Items: []v1beta2.Node: v1beta2.Node.Status: v1beta2.NodeStatus.Conditions: []v1beta2.Condition: decode slice: expect [ or n, but found {, error found in #10 byte of ...|ditions\":{},\"diskSta|..., bigger context ...|rCPURequest\":0,\"tags\":[]},\"status\":{\"conditions\":{},\"diskStatus\":{\"default-disk-745ff2be8d312b52\":{\"|..."
```

Reply:
Based on the info above, I think upgrading the API version then applying vxxxtoxxx upgrade would be more straightforward in terms of implementation.

Reply:
@shuo-wu Yes. But because of the CR status change, so we need to upgrade CR from v1beta1 to v1beta2 first, or we will hit the data type mismatch error.
I think this can be addressed in upgrade path refactoring task like https://github.com/longhorn/longhorn-manager/pull/1136#issuecomment-977630063